src/Parser.elm

```diff
@@ -0,0 +1,374 @@
+module Parser exposing
+    ( decodeAudiences
+    , decodeFolders
+    , decodeOneAudience
+    , decodeOneFolder
+    , parse
+    )
+
+import Data.Audience as A
+import Data.AudienceFolder as F
+import Dict
+import Json.Decode as Jd
+import Set
+
+
+parse : String -> String -> Result String Data
+parse rawFolders rawAudiences =
+    case ( parseFolders rawFolders, parseAudiences rawAudiences ) of
+        ( Err err, _ ) ->
+            Err <|
+                String.concat
+                    [ "error parsing folders JSON: "
+                    , Jd.errorToString err
+                    ]
+
+        ( _, Err err ) ->
+            Err <|
+                String.concat
+                    [ "error parsing audiences JSON: "
+                    , Jd.errorToString err
+                    ]
+
+        ( Ok folders, Ok audiences ) ->
+            let
+                clean =
+                    cleanUpData folders audiences
+            in
+            case checkData clean of
+                Nothing ->
+                    Ok clean
+
+                Just err ->
+                    Err err
+
+
+checkData : Data -> Maybe String
+checkData data =
+    maybeChain
+        [ allParentsAreFolders data
+        , allParentsExist data.all (Dict.toList data.parents)
+        ]
+
+
+allParentsExist : Dict.Dict Int String -> List ( Int, Int ) -> Maybe String
+allParentsExist all parents =
+    case parents of
+        [] ->
+            Nothing
+
+        ( child, parent ) :: arents ->
+            case Dict.get parent all of
+                Nothing ->
+                    Just <|
+                        String.concat
+                            [ "parent with ID "
+                            , String.fromInt parent
+                            , " and child "
+                            , String.fromInt child
+                            , " does not exist"
+                            ]
+
+                Just _ ->
+                    allParentsExist all arents
```

Comment:
There are nicer solutions possible than the explicit recursion:
* if you're OK with just a `Bool` output (no further data about what's missing, just that _something's_ missing) then `List.all (parentExist all) parents` would do just fine
* if you want all errors, you can `List.filterMap ... parents`

The current solution returns the first error but not all of them.

Reply:
I agree that mine is a bit ugly, but what I quite like is that it quits the map on the first error, which is what I want. Though I suppose it doesn't matter finding all of them; it might even help. In my new refactoring, I made a badMap function, for occasions when you want to quit the map when something goes wrong:

```
badMap : (a -> Result b c) -> List a -> Result b (List c)              
badMap f as_ =                                                         
    badMapHelp f as_ []                                                

                                                                       
badMapHelp : (a -> Result b c) -> List a -> List c -> Result b (List c)
badMapHelp f as_ accum =                                               
    case as_ of
        [] ->                                                          
            Ok (List.reverse accum)                                    

        a :: s ->                                                      
            case f a of
                Err err ->                                             
                    Err err                                            
    
                Ok ok ->                                               
                    badMapHelp f s (ok :: accum) 
```